src/allmydata/test/test_istorageserver.py

```diff
@@ -0,0 +1,282 @@
+
+    test_allocate_buckets_more_sharenums.skip = (
+        "https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3793"
+    )
+
+    @inlineCallbacks
+    def test_written_shares_are_allocated(self):
+        """
+        Shares that are fully written to show up as allocated in result from
+        ``IStoragServer.allocate_buckets()``.  Partially-written or empty
+        shares don't.
+        """
+        si, renew_secret, cancel_secret = (
+            new_storage_index(),
+            new_secret(),
+            new_secret(),
+        )
+        (_, allocated) = yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+
+        # Bucket 1 is fully written in one go.
+        yield allocated[1].callRemote("write", 0, b"1" * 1024)
+        yield allocated[1].callRemote("close")
+
+        # Bucket 2 is fully written in two steps.
+        yield allocated[2].callRemote("write", 0, b"1" * 512)
+        yield allocated[2].callRemote("write", 512, b"2" * 512)
+        yield allocated[2].callRemote("close")
+
+        # Bucket 0 has partial write.
+        yield allocated[0].callRemote("write", 0, b"1" * 512)
+
+        (already_got, _) = yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+        self.assertEqual(already_got, {1, 2})
+
+    @inlineCallbacks
+    def test_written_shares_are_readable(self):
+        """
+        Shares that are fully written to can be read.
+
+            1. The result is not affected by the order in which writes
+               happened, only by their offsets.
+
+            2. When overlapping writes happen, the resulting read returns the
+               earliest written value.
+        """
+        si, renew_secret, cancel_secret = (
+            new_storage_index(),
+            new_secret(),
+            new_secret(),
+        )
+        (_, allocated) = yield self.storage_server.allocate_buckets(
+            si,
+            renew_secret,
+            cancel_secret,
+            set(range(5)),
+            1024,
+            Referenceable(),
+        )
+
+        # Bucket 1 is fully written in order
+        yield allocated[1].callRemote("write", 0, b"1" * 512)
+        yield allocated[1].callRemote("write", 512, b"2" * 512)
+        yield allocated[1].callRemote("close")
+
+        # Bucket 2 is fully written in reverse.
+        yield allocated[2].callRemote("write", 512, b"4" * 512)
+        yield allocated[2].callRemote("write", 0, b"3" * 512)
+        yield allocated[2].callRemote("close")
+
+        # Bucket 3 has an overlapping write.
+        yield allocated[3].callRemote("write", 0, b"5" * 20)
+        yield allocated[3].callRemote("write", 0, b"5" * 24)
```

Comment:
Using `b"5"` for these first two writes means the test can't actually differentiate between keep-first and keep-second (for bytes 0 - 20) right?  Probably makes sense to use different values for the overwrite.